Remove `LazyArray` class from coords

dkillick opened the pull request:
In the past the coord `LazyArray` class was used to defer loading of all coords, but now it is only used to defer calculation of derived coordinates. However, there was still some `LazyArray` logic hanging around in the wider coords codebase, which was getting in the way of integrating the `DataManager` with coords. We have decided to remove the `LazyArray` class entirely to help this all along.

Note that doing this does not prevent derived coords from being calculated, but it does mean that the calculation is (currently) not deferred. When we come to [integrating derived coords with the `DataManager`](https://github.com/SciTools/iris/issues/2495) we can look to correcting this change, if by using dask deferral rather than by using the `LazyArray` class.

dkillick commented:
The four tests that were failing all expected that coords will be lazy. In the long term we want to maintain this behaviour, so I've introduced a temporary skip for these tests that can be removed when we reintroduce lazy coords via `dask` / `DataManager`.

lbdreyer commented on lib/iris/aux_factory.py:
```diff
@@ -1734,23 +1565,17 @@ def make_coord(self, coord_dims_func):
+            # Build the bounds array.
             nd_values_by_key = self._remap_with_bounds(dependency_dims,
                                                        derived_dims)
             bounds_shape = self._shape(nd_values_by_key)
```
It doesn't look like this gets used.

dkillick replied:
Whoops, that one must have slipped through. Good spot :+1:

lbdreyer commented on lib/iris/coords.py:
```diff
@@ -460,29 +460,16 @@ def __getitem__(self, key):
         """
+        points = self._points
+        bounds = self._bounds
+
```
Aren't these changes also in #2517 ? Why aren't there merge conflicts?

lbdreyer replied:
Okay there are actually merge conflicts, so this'll just need to be rebased.

dkillick replied:
Yeah, too many people all working in the same place...

pp-mo commented on lib/iris/aux_factory.py:
```diff
@@ -566,9 +449,7 @@ def calc_bounds():
-            b_shape = self._shape(nd_values_by_key)
-            b_dtype = self._dtype(nd_values_by_key)
-            bounds = _LazyArray(b_shape, calc_bounds, b_dtype)
+            bounds = calc_bounds()
```
I think it would be clearer here to ditch the `calc_bounds` function and use plain inline code, exactly as you did for `calc_points`.

dkillick replied:
@pp-mo I chose not to because the bounds calculation code is much longer and it was simpler to just leave the inline function and call it.

dkillick replied:
OK, maybe it isn't so bad pulling the code inline, but it still makes for a very long `make_coord` method. I wonder whether it would be better to pull the bounds calculation code up into a new method in each case, or whether to just leave that problem for the future. @pp-mo @lbdreyer thoughts?

pp-mo commented on lib/iris/aux_factory.py:
```diff
@@ -767,9 +640,7 @@ def calc_bounds():
-            b_shape = self._shape(nd_values_by_key)
-            b_dtype = self._dtype(nd_values_by_key)
-            bounds = _LazyArray(b_shape, calc_bounds, b_dtype)
+            bounds = calc_bounds()
```
Another vote for losing `calc_bounds`.

pp-mo replied:
P.S. (following offline discussion)
Changed my mind! We could instead leave all these one-use functions in place, to emphasize the common structure of these classes in support of one day simplifying + generalising..

pp-mo commented on lib/iris/aux_factory.py:
```diff
@@ -1004,12 +867,10 @@ def calc_bounds():
                                     nsigma_slice_bounds)
-
-            bounds = _LazyArray(bounds_shape, calc_bounds, bounds_dtype)
+            bounds = calc_bounds()
```
Another vote for losing `calc_bounds`.

dkillick commented:
I've rebased to sort out conflicts following the merge of #2517 and removed the extraneous line of code that @lbdreyer pointed out. 

Having discussed the bounds calculation enclosed function `calc_bounds` offline with @pp-mo we've agreed to keep things as they are. We decided this approach would make future work easier through making it clearer the extent of the bounds calculation functionality for each factory class.

lbdreyer commented:
Travis isn't running! Do you know why?

pp-mo commented on lib/iris/aux_factory.py:
```diff
@@ -957,27 +827,20 @@ def make_coord(self, coord_dims_func):
+            # Build the bounds array.
             nd_values_by_key = self._remap_with_bounds(dependency_dims,
                                                        derived_dims)
             bounds_shape = self._shape(nd_values_by_key)
```
This line is still redundant ...

dkillick replied:
It actually isn't... this is passed to `self._derive` on L871...

lbdreyer commented on lib/iris/tests/unit/fileformats/pyke_rules/compiled_krb/fc_rules_cf_fc/test_reorder_bounds_data.py:
```diff
@@ -56,19 +55,6 @@ def test_slowest_varying(self):
         expected = np.rollaxis(bounds_data, 0, bounds_data.ndim)
         self.assertArrayEqual(res, expected)
 
-    def test_slowest_varying_lazy(self):
```
Why did you choose to just remove this test rather than skip it like the others?

dkillick replied:
This test looked like it tested part of the implementation of `LazyArray`, rather than testing the general lazy coords concept. If you disagree or would like to retain the test then I can skip it instead.

lbdreyer replied:
It looks like the only places `reorder_bounds_data` is used is for building dimension coordinates and auxiliary coordinates at load time of netcdf files. Dimension coordinates doesn't use lazy data at this point and auxiliary coordinates uses a dask array.

pp-mo commented on lib/iris/aux_factory.py:
```diff
@@ -25,123 +25,14 @@
-from iris._deprecation import warn_deprecated
 from iris._cube_coord_common import CFVariableMixin
 import iris.coords
 import iris.util
```
This import is now obsolete.
